pkg/server/sotw/v3/server.go

```diff
@@ -183,35 +164,53 @@ func (s *server) process(str stream.Stream, reqCh <-chan *discovery.DiscoveryReq
 				}
 			}
 
-			if lastResponse, ok := lastDiscoveryResponses[req.TypeUrl]; ok {
-				if lastResponse.nonce == "" || lastResponse.nonce == nonce {
-					// Let's record Resource names that a client has received.
-					streamState.SetKnownResourceNames(req.TypeUrl, lastResponse.resources)
-				}
+			typeURL := req.GetTypeUrl()
+			// State cannot be modified until any potential watch is closed
+			state, ok := streamStates[typeURL]
+			if !ok {
+				// We don't have a current state for this type, create one
+				state = stream.NewStreamState(len(req.ResourceNames) == 0, nil)
+			} else if nonce != state.LastResponseNonce() {
+				// The request does not match the last response we sent.
+				// The protocol is a bit unclear in this case, but currently we discard such request and wait
+				// for the client to acknowledge the previous response.
+				// This is unclear how this handles cases where a response would be missed as any subsequent request will be discarded.
```

Comment:
The nonce is actually per type iiuc? That could be documented more clearly... The race that it exists to prevent only involve individual resource type. I think making the nonce per ADS stream (cross resource) would not work well: clients may not receive a newer discovery response but still be considered stale.

There is actually one case to at least take into consideration, explained here: https://github.com/envoyproxy/envoy/issues/10363.